src/System.Net.Http/src/System/Net/Http/HPack/DynamicTable.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace System.Net.Http.HPack
+{
+    internal class DynamicTable
+    {
+        private HeaderField[] _buffer;
+        private int _maxSize;
+        private int _size;
+        private int _count;
+        private int _insertIndex;
+        private int _removeIndex;
+
+        public DynamicTable(int maxSize)
+        {
+            _buffer = new HeaderField[maxSize / HeaderField.RfcOverhead];
+            _maxSize = maxSize;
+        }
+
+        public int Count => _count;
+
+        public int Size => _size;
+
+        public int MaxSize => _maxSize;
+
+        public HeaderField this[int index]
+        {
+            get
+            {
+                if (index >= _count)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                return _buffer[_insertIndex == 0 ? _buffer.Length - 1 : _insertIndex - index - 1];
+            }
+        }
+
+        public void Insert(Span<byte> name, Span<byte> value)
+        {
+            var entryLength = HeaderField.GetLength(name.Length, value.Length);
+            EnsureAvailable(entryLength);
+
+            if (entryLength > _maxSize)
+            {
+                // http://httpwg.org/specs/rfc7541.html#rfc.section.4.4
+                // It is not an error to attempt to add an entry that is larger than the maximum size;
+                // an attempt to add an entry larger than the maximum size causes the table to be emptied
+                // of all existing entries and results in an empty table.
+                return;
+            }
+
+            var entry = new HeaderField(name, value);
+            _buffer[_insertIndex] = entry;
+            _insertIndex = (_insertIndex + 1) % _buffer.Length;
+            _size += entry.Length;
+            _count++;
+        }
+
+        public void Resize(int maxSize)
+        {
+            if (maxSize > _maxSize)
+            {
+                var newBuffer = new HeaderField[maxSize / HeaderField.RfcOverhead];
```

Comment:
Would ther eben any benefit in growing to `Math.Max(maxSize, _maxSize * 2)`, or some such thing?

Reply:
I'm not sure. Actually I'm not entirely sure how we can get to this code in general. There's a connection level limit to the table size, and it looks like we initially allocate to that table size, so it's not clear how we'd ever get to a situation where the server can cause _maxSize to increase here. 

I'm going to add a TODO here and we can revisit later.